PM: Consider the time required to exit an idle state

ceolin opened the pull request:
- Add support for exit-latency-us on dts
   - This is the time required the CPU become ACTIVE from a given state
   - Wake up  the system this time earlier to be able to attend the scheduled event
- Remove some unnecessary code in samples

Commit:
dts: power: More information about min-residency-us

Make it clear that min-residency-us includes the time required to
entry in the state.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
dts: power: Add exit-latency-us field

Exit latency is the maximum time required by a CPU to transition from
a specific idle state to active state. This information is going to be
used to properly configure the wake up event in order to the system be
able to execute the next scheduled task.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
tests: power_states_api: Just use more realistic values

Using more realistic values to avoid confusion and make it clear the
distinction between min-residency-us and other idle properties like
exit-latency-us.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
power: Implement support for exit-latency-us

Add exit-latency-us in pm_state_info and extract this
information from the device tree.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
tests: power_states_api: Test exit-latency-us

Test exit-latency-us property on power API test.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
power: Consider the exit latency required by an idle state

Set the timeout to expire "exit_latency_us" earlier to the CPU be
capable of honor the next scheduled event.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
samples: ti: system_off: Remove unnecessary sleep

pm_power_state_force no longer requires the idle thread to run.  Just
removing the following k_sleep for two reasons, first it is not
necessary, second it tests that pm_power_state_force works as it
should.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
samples: nrf: system_off: Remove unnecessary sleep

pm_power_state_force no longer requires the idle thread to run.
Just removing the following k_sleep for two reasons, first it is
not necessary, second it tests that pm_power_state_force works
as it should.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

nashif commented on dts/bindings/power/state.yaml:
```diff
@@ -27,4 +27,5 @@ properties:
         description: |
             Minimum residency duration in microseconds. It is the minimum time for a
-            given idle state to be worthwhile energywise.
+            given idle state to be worthwhile energywise. It includes the time to enter
```
can we just take the description from Linux:

"Minimum residency duration in microseconds, inclusive of preparation
          and entry, for this idle state to be considered worthwhile energy wise"

nashif commented on dts/bindings/power/state.yaml:
```diff
@@ -28,4 +28,9 @@ properties:
+        type: int
+        required: false
+        description: |
+            Maximum time, in microseconds, required to wake up the CPU in this given state.
```
same, Linux:
"          Worst case latency in microseconds required to exit the idle state.
          The exit-latency-us duration may be guaranteed only after
          entry-latency-us has passed."

ceolin replied:
The last paragraph I have not included, we don't use/have entry latency.

nashif commented on include/power/power_state.h:
```diff
@@ -139,6 +141,14 @@ struct pm_state_info {
 	uint32_t min_residency_us;
+
+	/**
+	 * It is the maximum time in microseconds required to wake up the CPU
```
and then make this match the devicetree documnetation

nashif commented on subsys/power/power.c:
```diff
@@ -180,6 +181,15 @@ enum pm_state pm_system_suspend(int32_t ticks)
+	if (ticks != K_TICKS_FOREVER) {
+		/*
+		 * We need to set the timer to interrupt a little bit early to
+		 * accommodate the time required to the CPU fully wake up.
```
"..for the time required by the CPU to fully wake up."

erwango reviewed (commented):
Don't we need a check to ensure a minimum coherency between min-residency and exit-latency?
At least that the later is smaller than the former, to avoid unexpected behavior in kernel.

ceolin commented:
> Don't we need a check to ensure a minimum coherency between min-residency and exit-latency?
> At least that the later is smaller than the former, to avoid unexpected behavior in kernel.

I thought about it and decided to assume that the value set in dts is correct, we can do this basic check if min-residency is greater than exit-latency, but this is just a marginal test ... if the values are not correct it won't work properly. I can add an assert to catch this possible error though.

Commit:
power: policy: Account exit latency

Account exit latency time in the residency policy.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
power: Add an assert for an unexpected condition

Exit latency time should never be greater than the residency time.
Just add an assert in case the policy does not properly handle it.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

carlescufi commented:
@gmarull can you please take a look?

gmarull reviewed (approved):
LGTM
